### PR TITLE
Feature/exam multi correction/client refactoring

### PR DESCRIPTION
**Pull request by nicolasruscher**
### Checklist
- [x] I tested *all* changes and *all* related features with different users (student, tutor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.
- [x] Client: I followed the [coding and design guidelines](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/client.html).
- [x] Client: I documented the TypeScript code using JSDoc style.

### Motivation and Context
as a continuation to [https://github.com/ls1intum/Artemis/pull/2507](url), client changes still contained function calls like getLatestSubmissionResult() in HTML files.

### Description
Refactored getLatestSubmissionResult so the result of a submission is accessed via a map in the HTML files.

### Steps for Testing
Testing like in [https://github.com/ls1intum/Artemis/pull/2507 ](url), there shouldn't be any changes.

### CODE REVIEW

**Review by krusche — CHANGES_REQUESTED** [review]
I don't like the solution with Maps. It looks unnecessarily complex.
Why are you not simply storing the latestResult itself in the components?

Another alternativ would be to create a helper attribute in submission.model.ts called latestResult.

You would then simply set this on the submission itself. This way, you don't need a map

File: src/main/webapp/app/assessment/assessment-locks/assessment-locks.component.html [hunk 1]
```diff
@@ -37,7 +37,7 @@ <h2>
                     <td>{{ submission.submissionDate | date: 'MMM d, y HH:mm:ss' }}</td>
                     <td>{{ submission.participation.submissions?.length }}</td>
                     <td>
-                        <span *ngIf="getLatestSubmissionResult(submission)!.score != null">{{ getLatestSubmissionResult(submission)!.score }}%</span>
+                        <span *ngIf="submission.latestResult!.score != null">{{ submission.latestResult!.score }}%</span>
```
Comment (krusche) [inline src/main/webapp/app/assessment/assessment-locks/assessment-locks.component.html:40]: Please avoid the `!` operator. We never know if `latestResult` has a proper value.
Instead use the `?` operator so that it can fail.
Also avoid `null` and rather use `undefined` checks (I know this was here before, but we should change it)

Comment (krusche) [reply src/main/webapp/app/assessment/assessment-locks/assessment-locks.component.html:40]:
```suggestion
                        <span *ngIf="submission.latestResult?.score != undefined">{{ submission.latestResult.score }}%</span>
```

File: src/main/webapp/app/entities/submission.model.ts [hunk 2]
```diff
@@ -89,6 +91,12 @@ export function getFirstResult(submission: Submission | undefined): Result | und
     }
 }
 
+export function refreshLatestResult(submissions: Submission[]) {
```
Comment (krusche) [inline src/main/webapp/app/entities/submission.model.ts:94]:
```suggestion
export function setLatestResult(submissions: Submission[]) {
```

File: src/main/webapp/app/exercises/text/assess-new/conflicts/text-feedback-conflicts.component.ts [hunk 3]
```diff
@@ -121,8 +120,12 @@ export class TextFeedbackConflictsComponent extends TextAssessmentBaseComponent
 
         this.conflictingSubmissions = conflictingTextSubmissions;
         this.prepareTextBlocksAndFeedbackFor(this.leftSubmission!, this.leftTextBlockRefs, this.leftUnusedTextBlockRefs);
+        refreshLatestResult(conflictingTextSubmissions);
+        refreshLatestResult([this.leftSubmission!]);
         this.leftTotalScore = this.computeTotalScore(getLatestSubmissionResult(this.leftSubmission)!.feedbacks!);
         this.setConflictingSubmission(0);
+        refreshLatestResult(conflictingTextSubmissions);
```
Comment (krusche) [inline src/main/webapp/app/exercises/text/assess-new/conflicts/text-feedback-conflicts.component.ts:127]: why do we need it twice?

File: src/main/webapp/app/entities/submission.model.ts [hunk 4]
```diff
@@ -89,6 +91,12 @@ export function getFirstResult(submission: Submission | undefined): Result | und
...
+    submissions.forEach((submission) => {
+        submission.latestResult = getLatestSubmissionResult(submission);
+    });
+}
+
 // todo NR/SE: remove after PR can be merged into develop
```
Comment (krusche) [inline src/main/webapp/app/entities/submission.model.ts:100]: what about this remaining TODO?

File: src/main/webapp/app/entities/submission.model.ts (hunk 2 above)
Comment (krusche) [reply src/main/webapp/app/entities/submission.model.ts:94]: refresh sounds like retrieving it from the server, also make sure to add JavaDoc style documentation

Comment (krusche) [reply src/main/webapp/app/entities/submission.model.ts:94]: another thing: I would also offer this for one submission, then you don't need to wrap submissions in array when calling this method. As there is no overloading for Typescript functions, you need to use two different names

File: src/main/webapp/app/entities/submission.model.ts [hunk 5]
```diff
@@ -30,6 +30,8 @@ export abstract class Submission implements BaseEntity {
     public durationInMinutes?: number;
 
     public results?: Result[];
+    public latestResult?: Result;
```
Comment (krusche) [inline src/main/webapp/app/entities/submission.model.ts:33]: please add documentation what this is about and move it below into a new section 

`// Helper Attributes`

File: src/main/webapp/app/entities/submission.model.ts [hunk 6]
```diff
@@ -30,6 +30,8 @@ export abstract class Submission implements BaseEntity {
     public durationInMinutes?: number;
 
     public results?: Result[];
```
Comment (krusche) [inline src/main/webapp/app/entities/submission.model.ts:32]: also document this one as the attribute that is initialized if the server returns some values

File: src/main/webapp/app/entities/submission.model.ts (hunk 5 above)
Comment (krusche) [reply src/main/webapp/app/entities/submission.model.ts:33]: make sure to mention that it is undefined except the method below is invoked or it was explicitly set

**Review by krusche — CHANGES_REQUESTED** [review]
See my inline comments, in particular be careful with the `!` operator as it will lead to exception in case something is undefined. Try to be more failsafe and use the `?` operator.

In general, I would try to invoke the `refreshLatestResult` method (or the renamed version) directly in the service that loads the data from the server to move this logic away from the component which displays the data.

I added some other comments, please try to apply those for all the changes even if I just mention it in one case

File: src/main/webapp/app/exercises/text/assess-new/conflicts/text-feedback-conflicts.component.ts (hunk 3 above)
Comment (SimonEntholzer) [reply src/main/webapp/app/exercises/text/assess-new/conflicts/text-feedback-conflicts.component.ts:127]: During testing we discoverd that after `this.prepareTextBlocksAndFeedbackFor()` we need to refresh. Insiede of `this.setConflictingSubmission(0);` `this.prepareTextBlocksAndFeedbackFor()` is called again.
Since we have indiviudal refrehs now we can find out which are really necessary. I'll check it

File: src/main/webapp/app/entities/submission.model.ts (hunk 4 above)
Comment (SimonEntholzer) [reply src/main/webapp/app/entities/submission.model.ts:100]: I missed that one, I removed the funciton and its calls.

File: src/main/webapp/app/entities/submission.model.ts (hunk 6 above)
Comment (SimonEntholzer) [reply src/main/webapp/app/entities/submission.model.ts:33]: I added documentation to it

File: src/main/webapp/app/entities/submission.model.ts (hunk 5 above)
Comment (SimonEntholzer) [reply src/main/webapp/app/entities/submission.model.ts]: I added documentation and change latestResult to Result | undefined

File: src/main/webapp/app/exercises/text/assess-new/conflicts/text-feedback-conflicts.component.ts (hunk 3 above)
Comment (SimonEntholzer) [reply src/main/webapp/app/exercises/text/assess-new/conflicts/text-feedback-conflicts.component.ts]: Turns out we don't need it after all, it was only necessary when we still had the map for some reason. I removed the unused updates

File: src/main/resources/config/liquibase/master.xml [hunk 7]
```diff
@@ -94,5 +94,4 @@
     <include file="classpath:config/liquibase/changelog/20201111141743_changelog.xml" relativeToChangelogFile="false"/>
     <include file="classpath:config/liquibase/changelog/20201117113912_changelog.xml" relativeToChangelogFile="false"/>
     <include file="classpath:config/liquibase/changelog/20201119174900_changelog.xml" relativeToChangelogFile="false"/>
-    <include file="classpath:config/liquibase/changelog/20201214214200_changelog.xml" relativeToChangelogFile="false"/>
```
Comment (krusche) [inline src/main/resources/config/liquibase/master.xml:97]: please revert this change

Comment (SimonEntholzer) [reply src/main/resources/config/liquibase/master.xml]: I added it again

**Review by airelawaleria — DISMISSED** [review]
Tested in the testing session. Everything worked, approve

**Review by simonzilker — APPROVED** [review]
Tested in testing session. Everything works as intended!

**Review by EneaGore — DISMISSED** [review]
Tested independent of testing session, submissions for all exercises seem to work fine. Also tested Complaints and Feedback. Approve.

**Review by marvinchimeka — APPROVED** [review]
Everything works as intended. Approve

**Review by JohannesWeiss — APPROVED** [review]
Tested on ts1 as well, worked as intended.
When assessing one exercise after submitting the assessment, in the overview the amount did not increase (1/1), but went to (0/0) from (0/1). This has happened to me independent of this PR as well, therefore approve.
Screenshot 2020-12-19 at 21 18 25

File: src/main/webapp/app/entities/submission.model.ts [hunk 8]
```diff
@@ -62,36 +68,26 @@ export function getLatestSubmissionResult(submission: Submission | undefined): R
  * @param submission
  * @param result
  *
- * @return the index of the latestSubmissionResult or undefined if any of the params was undefined
+ * @return
```
Comment (nicolasruscher) [inline src/main/webapp/app/entities/submission.model.ts:71]: I think we can remove this, as the function returns nothing

File: src/main/webapp/app/entities/submission.model.ts [hunk 9]
```diff
@@ -29,9 +29,16 @@ export abstract class Submission implements BaseEntity {
...
     public durationInMinutes?: number;
 
+    // results is initialized by the value the server returns
     public results?: Result[];
+
     public participation?: Participation;
 
+    // Helper Attributes
+
+    // latestResult is undefined until submission.setLatestResult() is called
```
Comment (nicolasruscher) [inline src/main/webapp/app/entities/submission.model.ts:39]: its called setLatestSubmissionResult I think

File: src/main/webapp/app/entities/submission.model.ts [hunk 10]
```diff
@@ -62,36 +68,26 @@ export function getLatestSubmissionResult(submission: Submission | undefined): R
...
+        return;
     }
 
     if (submission!.results && submission!.results.length > 0) {
         submission.results[submission.results.length - 1] = result;
     } else {
         submission.results = [result];
     }
-    return submission.results.length - 1;
+    submission.latestResult = result;
```
Comment (nicolasruscher) [inline src/main/webapp/app/entities/submission.model.ts:83]: Quick Note, the method setLatestSubmissionResult isn't actually used in the code right now, at the moment we just do `submission.latestResult = getlatestSubmissionResult(submission);`
-> I think we have two option: remove the function, or rewrite the calls of  
`submission.latestResult = getlatestSubmissionResult(submission);` to 
`setLatestSubmissionResult(submission, getLatestSubmissionResult(submission))`

**Review by nicolasruscher — COMMENTED** [review]
Apart from the few comments it looks good, and much cleaner as before, good job! 👍

File: src/main/webapp/app/entities/submission.model.ts (hunk 9 above)
Comment (SimonEntholzer) [reply src/main/webapp/app/entities/submission.model.ts]: Yea, I changed the comment

File: src/main/webapp/app/entities/submission.model.ts (hunk 8 above)
Comment (SimonEntholzer) [reply src/main/webapp/app/entities/submission.model.ts]: yea, I fixed the doc

**Review by airelawaleria — APPROVED** [review]
Re-tested on ts1. Re-approve

**Review by simonzilker — DISMISSED** [review]
Tested in testing session. Everything worked fine!

**Review by JohannesWeiss — APPROVED** [review]
Tested in testing session on ts1 with all exercise types and exam combinations and worked as intended. Approve.

**Review by EneaGore — APPROVED** [review]
Tested in testing session, approve.

**Review by marvinchimeka — APPROVED** [review]
tested in a testing session on ts1. Re-Approve.

**Review by simonzilker — APPROVED** [review]
re-approve!

**Review by airelawaleria — DISMISSED** [review]
Re-approve

**Review by JohannesWeiss — APPROVED** [review]
Re-Approve!

**Review by EneaGore — DISMISSED** [review]
Re-approve

**Review by marvinchimeka — DISMISSED** [review]
Re-approve

**Review by airelawaleria — APPROVED** [review]
Re-approve

**Review by marvinchimeka — APPROVED** [review]
Re-approve

**Review by simonzilker — APPROVED** [review]
Re-approve

**Review by krusche — APPROVED** [review]
